backend/tests/controller/test_fhir_api_exception_handler.py

```diff
@@ -0,0 +1,75 @@
+import json
+import unittest
+from unittest.mock import patch
+
+from controller.fhir_api_exception_handler import fhir_api_exception_handler
+from models.errors import UnauthorizedError, UnauthorizedVaxError, ResourceNotFoundError
+
+
+class TestFhirApiExceptionHandler(unittest.TestCase):
+    def setUp(self):
+        self.logger_patcher = patch("controller.fhir_api_exception_handler.logger")
+        self.mock_logger = self.logger_patcher.start()
+
+    def tearDown(self):
+        patch.stopall()
+
+    def test_exception_handler_does_nothing_when_no_exception_occurs(self):
+        """Test that when the wrapped function returns successfully then the wrapper does nothing"""
+        @fhir_api_exception_handler
+        def dummy_func():
+            return "Hello World"
+
+        self.mock_logger.exception.assert_not_called()
+        self.assertEqual(dummy_func(), "Hello World")
+
+    def test_exception_handler_handles_custom_exception_and_returns_fhir_response(self):
+        """Test that custom exceptions are handled by the wrapper and a valid response is returned to the client"""
+        test_cases = {
+            ("Test UnauthorisedError", UnauthorizedError, 403, "forbidden", "Unauthorized request"),
+            ("Test UnauthorizedVaxError", UnauthorizedVaxError, 403, "forbidden",
+             "Unauthorized request for vaccine type"),
+            ("Test ResourceNotFoundError", ResourceNotFoundError, 404, "not-found", "Immunization resource does not "
+                                                                                    "exist. ID: 123")
+        }
+
+        for msg, error_type, expected_status, expected_code, expected_message in test_cases:
+            with self.subTest(msg=msg):
+
+                @fhir_api_exception_handler
+                def dummy_func():
+                    if msg == "Test ResourceNotFoundError":
+                        raise error_type(resource_type="Immunization", resource_id="123")
+
+                    raise error_type()
```

Comment:
Could provide the exact error to raise in the test case, then we wouldn't need the `msg` logic:
```suggestion
        test_cases = {
            (UnauthorizedError(), 403, "forbidden", "Unauthorized request"),
            (UnauthorizedVaxError(), 403, "forbidden", "Unauthorized request for vaccine type"),
            (
                ResourceNotFoundError(resource_type="Immunization", resource_id="123"),
                404,
                "not-found",
                "Immunization resource does not exist. ID: 123"
            )
        }

        for error, expected_status, expected_code, expected_message in test_cases:
            with self.subTest(msg=f`Test {error.__class__.__name__}`):

                @fhir_api_exception_handler
                def dummy_func():
                    raise error()
```

Reply:
Done - much nicer.